### ISSUE TITLE
 It shock me down! The gradient should be None but 0. it should be 0 but nan.

### ISSUE THREAD

**Issue by OuyangZhenqiu** [pytorch/pytorch#59172]
**The first question:**
with respect to the function, **tensor.backward()**, I caught up with a trouble. I look up the description of the function, 
**"Computes the gradient of current tensor w.r.t. graph leaves. The graph is differentiated using the chain rule."**
 
I am confused with graph leaves of the above-mentioned. Are they all graph leaves or the graph leaves which the tensor backwarding depends on?  

**If they are the latter,  the gradient should be None, instead of 0, of some parameters of mm independent on loss[1].
If the former, the gradient graph seems not to have any path from loss[1] to the parameters.** 

**Code:**
      xx = torch.tensor([1., 2., 3., 4.]) 
      target = torch.tensor([1., 2.]) 
      mm = torch.nn.Linear(4, 2)`
      loss_fn = nn.MSELoss(reduction="none")
      loss = loss_fn(mm(xx), target)
      for u in mm.parameters():
          print(u.grad)
      loss[1].backward()
      for u in mm.parameters():
          print(u.grad)
      loss = loss_fn(mm(xx), target)
      for u in mm.parameters():
          print(u.grad)
      loss[0].backward()
      for u in mm.parameters():
          print(u.grad)

**Result:**
      None
      None
      tensor([[ 0.0000,  0.0000,  0.0000,  0.0000],
              [-0.4103, -0.8206, -1.2308, -1.6411]])
      tensor([ 0.0000, -0.4103])
      tensor([[ 0.0000,  0.0000,  0.0000,  0.0000],
              [-0.4103, -0.8206, -1.2308, -1.6411]])
      tensor([ 0.0000, -0.4103])
      tensor([[ 1.2983,  2.5966,  3.8949,  5.1931],
              [-0.4103, -0.8206, -1.2308, -1.6411]])
      tensor([ 1.2983, -0.4103])

**The second question:**
The following thunder shock me down!  loss[0] is independent on  some parameters of mm. But  that it backwards leads to gradient of the parameters to become **nan**. That loss[1] is nan, but it did not backwards!  Compared to the first question,
the parameters depending loss[1], of mm, **become nan, instead of 0**.  I want to know what to lead the contradiction.
Could you help me settle the problem?

Code:
      xx = torch.tensor([1., 2., 3., 4.])
      target = torch.tensor([1., np.nan])
      mm = torch.nn.Linear(4, 2)
      loss_fn = nn.MSELoss(reduction="none")
      loss = loss_fn(mm(xx), target)
      for u in mm.parameters():
          print(u.grad)
      loss[0].backward()
      for u in mm.parameters():
          print(u.grad)`

      
Results:
      None
      None
      tensor([[ -9.1293, -18.2587, -27.3880, -36.5173],
              [     nan,      nan,      nan,      nan]])
      tensor([-9.1293,     nan])
Code:
      xx = torch.tensor([1., 2., 3., 4.])
      target = torch.tensor([1., np.nan])
      mm = torch.nn.Linear(4, 2)
      loss_fn = nn.MSELoss(reduction="none")
      loss = loss_fn(mm(xx), target)
      for u in mm.parameters():
          print(u.grad)
      loss[1].backward()
      for u in mm.parameters():
          print(u.grad)

Results:
      None
      None
      tensor([[0., 0., 0., 0.],
              [nan, nan, nan, nan]])
      tensor([0., nan])

**Comment by ngimel** [comment]
See related discussion in #15506

**Comment by OuyangZhenqiu** [comment]
> See related discussion in #15506
Hi! Thank you for your reply. In related discussion #15506, albanD said "it takes a big tensor as input and returns a smaller. So the backward takes a small tensor of gradients and returns a big tensor of gradients. a tensor can only contain numbers. And so the right value for the indexing op is to return 0 gradient in the places that were not used." 
I try to do an experiment to check whether it is right, and I find that the above explanation will not go down with me.
**In:
    a = torch.tensor([[1., 2, 3, 4],[1., 2, 3, 4]], requires_grad=True)
    b = torch.tensor([4, np.nan, 2, 1])
    c = a+b
    d = c[0, 0]
    d.backward()
    a.grad
Out:
tensor([[1., 0., 0., 0.],
        [0., 0., 0., 0.]])**
According to albanD's words,  the gradient of a[0,0] should be nan but 0. But If I replace operation '+' from "c = a + b" with '*',  there are as follows.
**Out:
tensor([[4., nan, 0., 0.],
        [0., nan, 0., 0.]])**
I hope you can help me to encounter the problem.

**Comment by albanD** [comment]
Hi,

That's because the `+` operation doesn't actually use the value of `b` to compute the gradients. So it is expected that you get `0` here.
As you have seen, if you use `*`, which will use the value of `b`, then you get nan values.

**Comment by OuyangZhenqiu** [comment]
> Hi,
> 
> That's because the `+` operation doesn't actually use the value of `b` to compute the gradients. So it is expected that you get `0` here.
> As you have seen, if you use `*`, which will use the value of `b`, then you get nan values.

Thank you. I understand.
